oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoVersionGCSupport.java

```diff
@@ -132,6 +133,56 @@ public CloseableIterable<NodeDocument> getPossiblyDeletedDocs(final long fromMod
                 input -> store.convertFromDBObject(NODES, input)));
     }
 
+    /**
+     * Calculate the bson representing including only the provided
+     * include path prefixes and/or excluding the provided
+     * exclude path prefixes - if any are provided - AND the provided
+     * query.
+     * Please note that at the moment the includes do not
+     * take long paths into account. That is, if a long path was
+     * supposed to be included via an include, it is not.
+     * Reason for this is that long paths would require
+     * the mongo query to include a '_path' condition - which disallows
+     * mongo from using the '_modified_id' index. IOW long paths
+     * would result in full scans - which results in bad performance.
+     * @param includes set of path prefixes which should only be considered
+     * @param excludes set of path prefixes which should be excluded.
+     * if these overlap with includes, then exclude has precedence.
+     * @param query the query with which to do an AND
+     * @return the combined bson with include/exclude path prefixes
+     * AND the provided query
+     */
+    private Bson withIncludeExcludes(Set<String> includes, Set<String> excludes, Bson query) {
+        Bson inclExcl = null;
+        if (includes != null && !includes.isEmpty()) {
+            final List<Bson> ors = new ArrayList<>(includes.size());
+            for (String incl : includes) {
+                ors.add(Filters.regex(ID, ":" + incl));
+            }
+            inclExcl = or(ors);
+        }
+        if (excludes != null && !excludes.isEmpty()) {
+            final List<Bson> ands = new ArrayList<>(excludes.size());
+            for (String excl : excludes) {
+                ands.add(not(Filters.regex(ID, ":" + excl)));
+            }
```

Comment:
Mongo will not use the index `_modified_id` if the query includes a negative condition. The reason is that `not(Filters.regex(ID, ":" + excl))` matches documents with `_id=null` and the indexes do not include null values. I had the same problem when implement include/exclude paths in the indexing job, the performance with include paths only was much better than when there were exclude paths. The fix I came up with is very ugly, but it works and provides a massive speed-up when downloading with excluded paths:

https://github.com/apache/jackrabbit-oak/blob/dbc728fc638fe1f6158165a23f24b418e200790e/oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/MongoDownloaderRegexUtils.java#L132-L141

https://issues.apache.org/jira/browse/OAK-10682

Reply:
Great, thanks for the heads-up @nfsantos ! I'll need to see how to fix this (worst case would I guess be to disallow exclude paths until that's improved)